File: core/rbac.py
"""RBAC — vai trò & quyền cho B2B."""

from __future__ import annotations

from typing import Literal

Role = Literal["admin", "analyst", "viewer"]

ROLES: tuple[Role, ...] = ("admin", "analyst", "viewer")

# Quyền → vai trò tối thiểu được phép
PERMISSIONS: dict[str, frozenset[Role]] = {
    "chat": frozenset({"admin", "analyst", "viewer"}),
    "explore": frozenset({"admin", "analyst", "viewer"}),
    "export": frozenset({"admin", "analyst", "viewer"}),
    "dashboard.read": frozenset({"admin", "analyst", "viewer"}),
    "dashboard.write": frozenset({"admin", "analyst"}),
    "article.write": frozenset({"admin", "analyst"}),
    "article.revise": frozenset({"admin", "analyst"}),
    "upload": frozenset({"admin", "analyst"}),
    "alerts.manage": frozenset({"admin", "analyst"}),
    "feedback": frozenset({"admin", "analyst", "viewer"}),
    "admin.tenants": frozenset({"admin"}),
    "admin.users": frozenset({"admin"}),
    "admin.keys": frozenset({"admin"}),
    "admin.branding": frozenset({"admin"}),
    "admin.jobs": frozenset({"admin"}),
}


def normalize_role(role: str | None) -> Role:
    r = (role or "viewer").strip().lower()
    if r in ROLES:
        return r  # type: ignore[return-value]
    return "viewer"


def has_permission(role: str | None, permission: str) -> bool:
    allowed = PERMISSIONS.get(permission)
    if allowed is None:
        return False
    return normalize_role(role) in allowed

File: web/src/lib/rbac.ts
/** RBAC phía client — đồng bộ tối thiểu với core/rbac.py */

export type ClientRole = "admin" | "analyst" | "viewer";

const PERMS: Record<string, ClientRole[]> = {
  chat: ["admin", "analyst", "viewer"],
  export: ["admin", "analyst"],
  "dashboard.write": ["admin", "analyst"],
  "article.write": ["admin", "analyst"],
  "article.revise": ["admin", "analyst"],
  upload: ["admin", "analyst"],
  "alerts.manage": ["admin", "analyst"],
  feedback: ["admin", "analyst", "viewer"],
  "admin.monitoring": ["admin"],
};

export function getClientRole(): ClientRole {
  if (typeof window === "undefined") return "analyst";
  const raw = (window.localStorage.getItem("abi_role") || "").toLowerCase();
  if (raw === "admin" || raw === "analyst" || raw === "viewer") return raw;
  // Chưa login (dev open) — cho phép analyst để không khóa demo
  if (!window.localStorage.getItem("abi_api_key")) return "analyst";
  return "viewer";
}

export function canClient(permission: string): boolean {
  const role = getClientRole();
  const allowed = PERMS[permission];
  if (!allowed) return role === "admin";
  return allowed.includes(role);
}
